### PR TITLE
BREAKING: Upgrade dependencies and implement new rules

### PR DESCRIPTION
**Pull request by MatthewHerbst**
Upgrades our ESLint peerDependency to its new major version:
`eslint` 3.19.0 -> 4.0.0

Upgrades two dependencies to their new minor versions:

`eslint-plugin-import` 2.2.0 -> 2.3.0
`eslint-plugin-react` 7.01 -> 7.1.0

With these upgrades, some new rules are available, and have
been implemented.

**Base config**
```
array-bracket-newline: 0
array-element-newline: 0
for-direction: 2
padding-line-between-statements: 0
semi-style: 2
switch-colon-spacing: 2
import/no-anonymous-default-export: 0
```

**React config**
```
react/no-redundant-should-component-update: 2
react/jsx-closing-tag-location: 2
```

Further, the rule `react/default-props-match-prop-types` should
now properly work - we seem to have preemptively added it before.

Finally, the rule `array-callback-return` was removed from the
recommended rules (which we currently still extend), and so was
placed into the full rule set.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -23,7 +23,10 @@ module.exports = {
   },
 
   rules: {
+    'array-bracket-newline': 0, // TODO: when supports single item line-breaked array (like `plugins` above): [2, { multiline: true }],
```
Comment (trotzig) [inline index.js]: The comment seems like an unfinished sentence to me. Is there a word missing?

File: index.js [hunk 2]
```diff
@@ -23,7 +23,10 @@ module.exports = {
...
     'array-bracket-spacing': [1, 'never'],
+    'array-callback-return': 2,
+    'array-element-newline': 0, // TODO: when supports single item line-breaked array (like `plugins` above): [2, { multiline: true }],
```
Comment (trotzig) [inline index.js]: Same here.

**Review by trotzig — APPROVED** [review]

File: index.js (hunk 1 above)
Comment (MatthewHerbst) [reply index.js]: I don't think so - what word should I add?

Comment (trotzig) [reply index.js]: "it" as in "when it supports ...". Or what are you trying to say here?

Comment (MatthewHerbst) [reply index.js]: It works :)

**Pull request merged**